File: LaboratoryWork/Semester_2/Laboratory_7/Laboratory_4/C++/task5/main.cpp
#include <iostream>
#include <random>
#include <map>
#include "strategies.h"

using namespace std;

typedef bool (*Strategy)(int, vector<bool>, vector<bool>);
const vector<vector<int>> POINTS = {
    {4, 20}, 
    {0, 24}
};

int main() {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> cntRounds(100, 200);
    uniform_int_distribution<> algorithmNum(0, 2);

    map<int, string> algorithmNames = {
        {0, "every23"},
        {1, "grimTrigger"},
        {2, "titForTat"}
    };

    int firstAlgorithm = algorithmNum(gen);
    int secondAlgorithm = algorithmNum(gen);
    while (firstAlgorithm == secondAlgorithm) {
        secondAlgorithm = algorithmNum(gen);
    }

    int rounds = cntRounds(gen);
    
    cout << "Rounds: " << rounds << endl << endl;
    cout << "First algorithm: " << algorithmNames[firstAlgorithm] << endl;
    cout << "Second algorithm: " << algorithmNames[secondAlgorithm] << endl << endl;

    Strategy strategies[] = {
        every23,
        grimTrigger,
        titForTat
    };

    vector<bool> historyFirst, historySecond;

    int pointsFirst = 0;
    int pointsSecond = 0;
    for (int i = 0; i < rounds; i++) {
        bool moveFirst;
        bool moveSecond;

        moveFirst = strategies[firstAlgorithm](i + 1, historyFirst, historySecond);
        moveSecond = strategies[secondAlgorithm](i + 1, historySecond, historyFirst);

        historyFirst.push_back(moveFirst);
        historySecond.push_back(moveSecond);

        pointsFirst += POINTS[moveFirst][moveSecond];
        pointsSecond += POINTS[moveSecond][moveFirst];
    }

    if (pointsFirst > pointsSecond) {
        cout << "First algorithm win!" << endl;
    } else if (pointsFirst < pointsSecond) {
        cout << "Second algorithm wins!" << endl;
    } else {
        cout << "It's a draw!" << endl;
    }

    cout << "First algorithm points: " << pointsFirst << endl;
    cout << "Second algorithm points: " << pointsSecond << endl;
}

File: LaboratoryWork/Semester_2/Laboratory_7/Laboratory_4/GoLang/task5/main.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

var POINTS = [][]int{
	{4, 20},
	{0, 24},
}

type Strategy func(int, []bool, []bool) bool

func main() {
	rand.Seed(time.Now().UnixNano())
	cntRounds := rand.Intn(101) + 100
	algorithmNum := func() int {
		return rand.Intn(3)
	}

	algorithmNames := map[int]string{
		0: "every23",
		1: "grimTrigger",
		2: "titForTat",
	}

	firstAlgorithm := algorithmNum()
	secondAlgorithm := algorithmNum()
	for firstAlgorithm == secondAlgorithm {
		secondAlgorithm = algorithmNum()
	}

	fmt.Println("Rounds:", cntRounds)
	fmt.Println()
	fmt.Println("First algorithm:", algorithmNames[firstAlgorithm])
	fmt.Println("Second algorithm:", algorithmNames[secondAlgorithm])
	fmt.Println()

	strategies := []Strategy{
		every23,
		grimTrigger,
		titForTat,
	}

	historyFirst := []bool{}
	historySecond := []bool{}

	pointsFirst := 0
	pointsSecond := 0

	for i := 0; i < cntRounds; i++ {
		moveFirst := strategies[firstAlgorithm](i+1, historyFirst, historySecond)
		moveSecond := strategies[secondAlgorithm](i+1, historySecond, historyFirst)

		historyFirst = append(historyFirst, moveFirst)
		historySecond = append(historySecond, moveSecond)

		var valFirst, valSecond int
		if moveFirst {
			if moveSecond {
				valFirst = POINTS[1][1]
				valSecond = POINTS[1][1]
			} else {
				valFirst = POINTS[1][0]
				valSecond = POINTS[0][1]
			}
		} else {
			if moveSecond {
				valFirst = POINTS[0][1]
				valSecond = POINTS[1][0]
			} else {
				valFirst = POINTS[0][0]
				valSecond = POINTS[0][0]
			}
		}

		pointsFirst += valFirst
		pointsSecond += valSecond
	}

	if pointsFirst > pointsSecond {
		fmt.Println("First algorithm win!")
	} else if pointsFirst < pointsSecond {
		fmt.Println("Second algorithm wins!")
	} else {
		fmt.Println("It's a draw!")
	}

	fmt.Println("First algorithm points:", pointsFirst)
	fmt.Println("Second algorithm points:", pointsSecond)
}
